### PR TITLE
orchestrator upgrade: use ContainerService's OrchestratorProfile as a GoalState

### PR DESCRIPTION
**Pull request by dmitsh**
- use ContainerService's OrchestratorProfile as a GoalState 
 - add UpgradeContainerService to versioned and unversioned APIs
 - eliminate duplicated structures

### CODE REVIEW

File: pkg/api/types.go [hunk 1]
```diff
@@ -55,6 +55,7 @@ type Properties struct {
 	AADProfile              *AADProfile              `json:"aadProfile,omitempty"`
 	CustomProfile           *CustomProfile           `json:"customProfile,omitempty"`
 	HostedMasterProfile     *HostedMasterProfile     `json:"hostedMasterProfile,omitempty"`
+	UpgradeProfile          *UpgradeContainerService `json:"upgradeProfile,omitempty"`
```
Comment (amanohar) [inline pkg/api/types.go]: I don't understand the purpose of this change.  Plus it's not the correct way to represent old and new versions. Can you elaborate?

Comment (dmitsh) [reply pkg/api/types.go]: reverted as discussed

File: pkg/api/converterfromupgradeapi.go [hunk 2]
```diff
@@ -10,18 +13,38 @@ import "github.com/Azure/acs-engine/pkg/api/vlabs"
...
 ///////////////////////////////////////////////////////////
 
-// ConvertUpgradeContainerServiceToVLabs converts an unversioned ContainerService to a vlabs ContainerService
-func ConvertUpgradeContainerServiceToVLabs(api *UpgradeContainerService) *vlabs.UpgradeContainerService {
-	vlabsUCS := &vlabs.UpgradeContainerService{}
-	vlabsUCS.OrchestratorProfile = &vlabs.OrchestratorProfile{}
-	convertUpgradeOrchestratorProfileToVLabs(api.OrchestratorProfile, vlabsUCS.OrchestratorProfile)
-	return vlabsUCS
+// ConvertUpgradeContainerServiceToV20170930 converts an unversioned UpgradeContainerService to a v20170930 UpgradeContainerService
+func ConvertUpgradeContainerServiceToV20170930(api *UpgradeContainerService) *v20170930.UpgradeContainerService {
```
Comment (amanohar) [inline pkg/api/converterfromupgradeapi.go]: Based on previous discussion wasn't this going to be removed?

File: pkg/operations/kubernetesupgrade/upgradecluster.go [hunk 3]
```diff
@@ -152,6 +147,39 @@ func (uc *UpgradeCluster) getClusterNodeStatus(subscriptionID uuid.UUID, resourc
...
+		return fmt.Errorf("Unsupported orchestrator tag format %s", vmOrchestratorTypeAndVersion)
+	}
+	currentVer := arr[1]
+	arr = strings.Split(currentVer, ".")
+	if len(arr) != 3 {
+		return fmt.Errorf("Unsupported orchestrator version format %s", currentVer)
+	}
+	currentRel := fmt.Sprintf("%s.%s", arr[0], arr[1])
+
+	cs := &api.ContainerService{
```
Comment (amanohar) [inline pkg/operations/kubernetesupgrade/upgradecluster.go]: Why does it need to construct a whole new ContainerService model to get GetOrchestratorVersionProfile

Comment (dmitsh) [reply pkg/operations/kubernetesupgrade/upgradecluster.go]: addressed

File: pkg/api/converterfromupgradeapi.go (hunk 2 above)
Comment (dmitsh) [reply pkg/api/converterfromupgradeapi.go]: addressed

File: cmd/upgrade.go [hunk 4]
```diff
@@ -110,19 +109,22 @@ func (uc *upgradeCmd) validate(cmd *cobra.Command, args []string) {
...
 	}
-	uc.upgradeContainerService, uc.upgradeAPIVersion, err = upgradeapiloader.LoadUpgradeContainerServiceFromFile(uc.upgradeModelFile)
+	upgradeModel, err := apiloader.LoadUpgradeContainerService(contents, vlabs.APIVersion, true)
 	if err != nil {
-		log.Fatalf("error parsing the upgrade api model: %s", err.Error())
+		log.Fatalf("error loading UpgradeContainerService: %s", err.Error())
 	}
+	// set GoalState
+	uc.containerService.Properties.OrchestratorProfile.OrchestratorRelease = upgradeModel.OrchestratorRelease
+	uc.containerService.Properties.OrchestratorProfile.OrchestratorVersion = upgradeModel.OrchestratorVersion
```
Comment (amanohar) [inline cmd/upgrade.go]: IIRC the decision was that customer was only going to provide in the input OrchestratorRelease and OrchestratorVersion  would be determined by the upgrade operation.

File: cmd/upgrade.go [hunk 5]
```diff
@@ -110,19 +109,26 @@ func (uc *upgradeCmd) validate(cmd *cobra.Command, args []string) {
...
+		log.Fatalf("error reading file %s: %s", uc.upgradeModelFile, err.Error())
 	}
-	uc.upgradeContainerService, uc.upgradeAPIVersion, err = upgradeapiloader.LoadUpgradeContainerServiceFromFile(uc.upgradeModelFile)
+	upgradeModel, err := apiloader.LoadUpgradeContainerService(contents, vlabs.APIVersion, true)
 	if err != nil {
-		log.Fatalf("error parsing the upgrade api model: %s", err.Error())
+		log.Fatalf("error loading UpgradeContainerService: %s", err.Error())
+	}
+	// validate desired release
+	if err = kubernetesupgrade.PreValidateUpgrade(uc.containerService, upgradeModel, true); err != nil {
```
Comment (amanohar) [inline cmd/upgrade.go]: As discussed the internal upgrade model is not needed. Convert directly to api.containerservice type after validation.

**Pull request merged**